docs/source/pynq_libraries/pynqmb_reference.rst

```diff
@@ -0,0 +1,252 @@
+are in ``i2c.h``.
+
+``i2c type``
+~~~~~~~~~~~~
+
+Represents an I2C master. It is possible for multiple handles to
+reference the same master device.
+
+``i2c i2c_open(int sda, int scl)``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Open an I2C device attached to an I/O switch configured to use the
+specified pins. Calling this function will disconnect any previously
+assigned pins and return them to a high-impedance state.
+
+``i2c i2c_open_device(unsigned int device)``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Open an I2C master by base address or ID
+
+``void i2c_read(i2c dev_id, unsigned int slave_address, unsigned char* buffer, unsigned int length)``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Issue a read command to the specified slave. ``buffer`` is an array
+allocated by the caller of at least length ``length``.
+
+``void i2c_write(i2c dev_id, unsigned int slave_address, unsigned char* buffer, unsigned int length)``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Issue a write command to the specified slave.
+
+``void i2c_close(i2c dev_id)``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Close the I2C device.
+
+SPI Devices
+~~~~~~~~~~~
+
+SPI operates on a synchronous transfer of data so rather than read and
+write, only a ``transfer`` function is provided. These functions are all
+provided by ``spi.h``.
+
+``spi`` type
+~~~~~~~~~~~~
+
+Handle to a SPI master.
+
+``spi spi_open(unsigned int spiclk, unsigned int miso, unsigned int mosi, unsigned int ss)``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Opens a SPI master on the specified pins. If a pin is not needed for a
+device, ``-1`` can be passed in to leave it unconnected.
+
+``spi spi_open_device(unsigned int device)``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Opens a SPI master by base address or device ID.
+
+``spi spi_configure(spi dev_id, unsigned int clk_phase, unsigned int clk_polarity)``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Configures the SPI master with the specified clock phase and polarity.
+These settings are global to all handles to a SPI master.
+
+\`void spi\_transfer(spi dev\_id, const char\* write\_data, char\* read\_data, unsigned int length);
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Transfer bytes to and from the SPI slave. Both ``write_data`` and
+``write_data`` should be allocated by the caller and NULL. Buffers
+should be at least of length ``length``.
+
+``void spi_close(spi dev_id)``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Closes a SPI master
+
+Timer Devices
+-------------
+
+Timer devices serve two purposes. They can either be used to output PWM
+signals or as program timers for inserting accurate delays. All of these
+functions are in ``timer.h``
```

Comment:
should we also mention that if a timer is used for `delay_ms` and such, it cannot be used for other functionalities?

Reply:
Yes we should